Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy.swift

```diff
@@ -460,3 +462,72 @@ extension PathHierarchy.DisambiguationContainer {
         }))
     }
 }
+
+// MARK: Deserialization
+
+extension PathHierarchy {
+    // This is defined in the main PathHierarchy.swift file to access fileprivate properties and PathHierarchy.Node API without making it internally visible.
+    
+    // This mapping closure exist so that we don't encode ResolvedIdentifier values into the file. They're an implementation detail and they are a not stable across executions.
+    
+    /// Decode a path hierarchy from its file representation.
+    ///
+    /// The caller can use `mapCreatedIdentifiers` when encoding and decoding path hierarchies to associate auxiliary data with a node in the hierarchy.
+    ///
+    /// - Parameters:
+    ///   - fileRepresentation: A file representation to decode.
+    ///   - mapCreatedIdentifiers: A closure that the caller can use to map indices to resolved identifiers.
+    init(
+        _ fileRepresentation: FileRepresentation,
+        mapCreatedIdentifiers: (_ identifiers: [ResolvedIdentifier]) -> Void
+    ) {
+        // Generate new identifiers. While building the path hierarchy, the node numbers map to identifiers via index lookup in this array.
+        var identifiers = [ResolvedIdentifier]()
+        identifiers.reserveCapacity(fileRepresentation.nodes.count)
+        for _ in 0 ..< fileRepresentation.nodes.count {
+            identifiers.append(ResolvedIdentifier())
+        }
+        
+        var lookup = [ResolvedIdentifier: Node]()
+        lookup.reserveCapacity(fileRepresentation.nodes.count)
+        // Iterate once to create all the nodes
+        for (index, fileNode) in zip(0..., fileRepresentation.nodes) {
+            let node: Node
+            if let symbolID = fileNode.symbolID {
+                // Symbols decoded from a file representation only need an accurate ID. The rest of the information is never read and can be left empty.
+                let symbol = SymbolGraph.Symbol(
+                    identifier: symbolID,
+                    names: .init(title: "", navigator: nil, subHeading: nil, prose: nil),
+                    pathComponents: [],
+                    docComment: nil,
+                    accessLevel: .public,
+                    kind: SymbolGraph.Symbol.Kind(rawIdentifier: "", displayName: ""),
+                    mixins: [:]
+                )
+                node = Node(symbol: symbol, name: fileNode.name)
+            } else {
+                node = Node(name: fileNode.name)
+            }
+            node.isDisfavoredInCollision = fileNode.isDisfavoredInCollision
+            node.identifier = identifiers[index]
+            lookup[node.identifier] = node
+        }
+        // Iterate again to construct the tree
+        for (index, fileNode) in zip(0..., fileRepresentation.nodes) {
```

Comment:
Would `.enumerated()` work here? 

```suggestion
        for (index, fileNode) in fileRepresentation.nodes.enumerated() {
```

Reply:
Yes, using `.enumerated()` would be correct because it's indexing from 0.

That said, I have an irrational dislike for the `enumerated()` API because it looks like the enumeration number can be used to subscript into the collection but that isn't generally true. For example, when dropping an elements from an array, the first index of the returned array slice is offset but enumerated always starts at zero and using it to subscript is out of bounds in the slice. Because of this I personally prefer to be explicit about when I'm using `0` as a starting index rather than getting that behavior implicitly from `enumerated()`.